src/lib.rs

```diff
@@ -358,6 +363,175 @@ fn jpeg_size<R: BufRead + Seek>(reader: &mut R) -> ImageResult<ImageSize> {
     })
 }
 
+fn jxl_size<R: BufRead + Seek>(reader: &mut R) -> ImageResult<ImageSize> {
+    let mut file_header = [0; 16]; // The size is variable, but doesn't exceed 16 bytes
+    let mut header_size = 0;
+
+    reader.seek(SeekFrom::Start(0))?;
+    reader.read_exact(&mut file_header[..2])?;
+
+    if &file_header[..2] == b"\xFF\x0A" {
+        // Raw data: Read header directly
+        header_size = reader.read(&mut file_header[2..])? + 2;
+    } else {
+        // Container format: Read from a single jxlc box or multiple jxlp boxes
+        reader.seek(SeekFrom::Start(12))?;
+
+        loop {
+            let (box_type, box_size) = next_tag(reader)?;
+            let box_start = reader.stream_position()? - 8;
+
+            // If box_size is 1, the real size is stored in the first 8 bytes of content.
+            // If box_size is 0, the box ends at EOF.
+
+            let box_size = match box_size {
+                1 => {
+                    let mut box_size = [0; 8];
+                    reader.read_exact(&mut box_size)?;
+                    u64::from_be_bytes(box_size)
+                },
+                _ => box_size as u64,
+            };
+
+            let box_end = box_start.checked_add(box_size).ok_or(ImageError::CorruptedImage)?;
+            let box_header_size = reader.stream_position()? - box_start;
+
+            if box_size < box_header_size && box_size != 0 {
+                return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("Invalid size for {} box: {}", box_type, box_size)).into());
+            }
+
+            // The jxlc box must contain the complete codestream
+
+            if box_type == "jxlc" {
+                let read_end = match box_size {
+                    0 => 16,
+                    _ => std::cmp::min(box_size - box_header_size, 16) as usize,
+                };
+
+                header_size = reader.read(&mut file_header[..read_end])?;
+                break;
+            }
+
+            // Or it could be stored as part of multiple jxlp boxes
+
+            if box_type == "jxlp" {
+                let mut jxlp_index = [0; 4];
+                reader.read_exact(&mut jxlp_index)?;
+
+                if box_size - box_header_size < 4 && box_size != 0 {
```

Comment:
On this line there's an issue with subtraction underflow in debug mode. There is a check for `box_size >= box_header_size` above, so it may be sufficient to just flip the checks here so the 0 is accounted for first.

Reply:
I updated that part so it uses a Take adaptor instead, now it doesn't need those explicit checks and it also solves the panic you mentioned in your other comment.